fake-lmp-device-register

```diff
@@ -29,43 +29,62 @@ class Options(NamedTuple):
 
 
 def create_key(uuid: str, factory: str, production: bool) -> Tuple[bytes, bytes]:
-    key = subprocess.check_output(
-        ["openssl", "ecparam", "-genkey", "-name", "prime256v1"]
-    )
-
-    with NamedTemporaryFile() as cnf:
-        cnf.write(
-            f"""[req]
-prompt = no
-distinguished_name = dn
-req_extensions = ext
-
-[dn]
-CN={uuid}
-OU={factory}
-""".encode()
-        )
-        if production:
-            cnf.write(b"businessCategory=production\n")
-        cnf.write(
-            b"""
-[ext]
-keyUsage=critical, digitalSignature
-extendedKeyUsage=critical, clientAuth
-"""
-        )
-        cnf.flush()
-
-        r = subprocess.run(
-            ["openssl", "req", "-new", "-config", cnf.name, "-key", "-"],
-            input=key,
-            stdout=subprocess.PIPE,
-        )
-        r.check_returncode()
+    try:
+        log.info("Creating key for UUID: %s, Factory: %s", uuid, factory)
+
+        # Create a private key and store it in a temp file
+        with NamedTemporaryFile(delete=False, suffix=".pem") as keyfile:
+            key = subprocess.check_output(["openssl", "ecparam", "-genkey", "-name", "prime256v1"])
+            keyfile.write(key)
+            keyfile_path = keyfile.name
+            keyfile.flush()
+
+        # Create a temp config for OpenSSL's CSR generation
+        with NamedTemporaryFile(delete=False) as cnf:
+            cnf.write(f"""[req]
+    prompt = no
+    distinguished_name = dn
+    req_extensions = ext
+
+    [dn]
+    CN={uuid}
+    OU={factory}
+    """.encode()
+            )
+            if production:
+                cnf.write(b"businessCategory=production\n")
+            cnf.write(
+                b"""
+    [ext]
+    keyUsage=critical, digitalSignature
+    extendedKeyUsage=critical, clientAuth
+    """
+            )
+            cnf.flush()
+            cnf.close()
+
+            # Use OpenSSL to generate a CSR using the private key and config
+            r = subprocess.run(["openssl", "req", "-new", "-config", cnf.name, "-key", keyfile_path], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+            if r.returncode != 0:
+                log.error("Error in running openssl command:")
+                log.error(r.stderr.decode())
+
+                # Clean up temp files in case of errors
+                os.unlink(cnf.name)
+                os.unlink(keyfile_path)
+                raise subprocess.CalledProcessError(r.returncode, r.args)
+
+    except Exception as e:
+        log.error("An error occurred: %s", e)
+    finally:
+        # Let's ensure that temp files are deleted
+        os.unlink(cnf.name)
+        os.unlink(keyfile_path)
```

Comment:
I think you can simplify this quite a bit by using both tempfile in the same context. e.g:
```
keyfile = NamedTemporaryFile()
cnf = NamedTemporaryFile()
with keyfile, cnf:
    # key file logic
    keyfile.flush()
    # create the csr
```

Reply:
Just IMHO: we don't need the key file to be a temp file, because the next thing we do at line 97 is... save the key file to `opts.sota_dir`.

So, I'd restructure this code to do the following:
1. Save the cnf into a tmp file.
2. Generate the key file into `{opts.sota_dir}/pkey.pem`.
3. Generate csr into stdout (as is now) and return it.

That should save a lot of tmpfile and file.flush operations.
Moreover, it would allow to keep the original code structure to the max, as we'd only need to change two operations - to generate the key, and to use the key.

Reply:
+1 The suggestion was applied